Source/Core/VideoCommon/PixelShaderGen.cpp

```diff
@@ -567,7 +567,42 @@ int4 readTextureLinear(in Texture2DArray tex, uint2 uv1, uint2 uv2, int layer, i
 }}
 )");
 
-    out.Write(R"(
+    if (host_config.manual_texture_sampling_custom_texture_sizes)
+    {
+      // This is slower, and doesn't result in the same odd behavior that happens on console when
+      // wrapping with non-power-of-2 sizes, but it's fine for custom textures to have non-console
+      // behavior.
+      out.Write(R"(
+// Both GLSL and HLSL produce undefined values when the modulo operator (%) is used with a negative
+// dividend and a positive divisor.  We want a positive value such that FlooredModulo(-1, 3) is 2.
+// We don't really care about what happens when the divisor is negative, so Euclidean division would
+// also work, but floored division is easier to implement.  Note that when both the dividend and
+// divisor are positive, FlooredModulo(dividend, divisor) is euqal to dividend % divisor.
+// See https://en.wikipedia.org/wiki/Modulo_operation#Variants_of_the_definition
+int FlooredModulo(int dividend, int divisor) {{
+  return dividend - divisor * int(floor(float(dividend) / divisor));
+}}
+
+uint WrapCoord(int coord, uint wrap, int size) {{
+  switch (wrap) {{
+    case {:s}:
+    default: // confirmed that clamp is used for invalid (3) via hardware test
+      return uint(clamp(coord, 0, size - 1));
+    case {:s}:
+      return uint(FlooredModulo(coord, size));  // coord % size
```

Comment:
Should this be:

```
return uint(FlooredModulo(coord, size - 1));
```

?

Reply:
No, it is correct; `FlooredModulo` works in the half-open interval `[0, size)` (i.e. the closed interval `[0, size - 1]`) already.